pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -81,126 +62,114 @@ func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
 	}
 
+	conn, err := existingPlugin.getOrCreateGRPCConn()
+	if err != nil {
+		return nil, err
+	}
+
 	return &draPluginClient{
-		pluginName:        pluginName,
-		addr:              draAddr(existingPlugin.endpoint),
-		nodeClientCreator: newNodeClient,
+		pluginName:       pluginName,
+		plugin:           existingPlugin,
+		pluginAPIVersion: detectVersion(conn),
 	}, nil
 }
 
 func (r *draPluginClient) NodePrepareResources(
 	ctx context.Context,
-	req *drapb.NodePrepareResourcesRequest,
+	req *drapbv1alpha3.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (response *drapbv1alpha3.NodePrepareResourcesResponse, err error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
-
-	if r.nodeClientCreator == nil {
-		return nil, errors.New("failed to call NodePrepareResources. nodeClientCreator is nil")
-	}
+	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", response, "err", err)
 
-	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
+	conn, err := r.plugin.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	defer closer.Close()
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	resp, err = nodeClient.NodePrepareResources(ctx, req)
-	if err != nil {
-		status, _ := grpcstatus.FromError(err)
-		if status.Code() == grpccodes.Unimplemented {
-			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
+	if r.pluginAPIVersion == "v1alpha3" {
+		nodeClient := drapbv1alpha3.NewNodeClient(conn)
+		response, err = nodeClient.NodePrepareResources(ctx, req)
+		return
```

Comment:
We should explicitly return the values here, not rely on the named return values. We should also wrap the error with something like:
```
if err != nil {
    return nil, fmt.Errorf("error calling v1alpha3.NodePrepareResources: %v", err)
}
return response, err
```

Reply:
done in 38653c477e7590234a92dc58d9fb73a1c7a6f385